fuse_core/src/timestamp_manager.cpp

```diff
@@ -156,6 +156,11 @@ void TimestampManager::query(
     // Insert the last timestamp into the motion model history, but with no constraints. The last entry in the motion
     // model history will always contain no constraints.
     motion_model_history_.emplace(last_stamp, MotionModelSegment());
+
+    // Call the motion model generator so it inserts the last timestamp into its state history.
```

Comment:
I wonder if something cleaner could be done, but I guess for that we'll have to make this class know about the motion model. :thinking:

Reply:
Here I just want to hear your thoughts, but this is certainly something for a another PR, or we just go with this for now.

Reply:
Admittedly, I'm not in love with calling the motion model generator with these degenerate time ranges. Particularly since this is only an issue for initializing the first state, but the solution adds an additional call to every transaction.

One option I thought about was modifying the motion model implementation here:
https://github.com/locusrobotics/fuse/blob/devel/fuse_models/src/unicycle_2d.cpp#L106

At this point, we have access to the input transaction with all of the timestamps and added variables. This would allow us to steal the initial values of the variables added by the ignition sensor, instead of using this call with a degenerate time range to insert the initial state. The implications of that get tangled. In theory, that would mean the motion model could initialize itself properly from the very beginning, and does not need a round trip to the optimizer and back to work correctly. Thus, the requirement to process the ignition constraint in its own transaction falls away. But if we relax that requirement, we are relying on the incoming transaction to have valid initial variable values. Right now, when transactions are combined, the variable values from the first transaction are retained. So as long as the first transaction is the ignition transaction, then that should work. But it felt a bit brittle. It also means we are moving coding complexity out of the Optimizer class and into the MotionModel class. And I wasn't sure that was a good trade.

Regardless, I would want to test that option more before I decided it was a good idea. And it could be that processing the ignition sensor in a separate transaction along with modifications to the motion model would be a good combination.

This current PR fixes the issue now, so I'm all for getting it merged. We can always cahnge things up later.

Reply:
I'm 99% this is not only an issue for initializing the first state, if you mean the very first entry in the history. With relative constraints, the second variable in the constraint is usually (if not always) entering here, and therefore it was also impacted. I might be wrong though, because I was printing things here and getting a ton of spam. A unit test that exercise that would help. Not sure if you already cover that case in your unit test PR.

Anyway, I think we can merge this an continue the discussion in https://github.com/locusrobotics/fuse/pull/158

Reply:
At least with my understanding of the problem, I only see two conditions that cause problems:
1. If the state history is empty, then the lookup into the state history for the first involved timestamp fails. When that fails, it uses zeros as the initial state.
2. If a constraint arrives out-of-order and/or references a timestamp before the beginning of the state history, then the lookup into the state history for that timestamp will also fail and result in a zero state estimate for that state.

Generally speaking, relative pose constraints should work just fine after the state history has been properly initialized. The motion model will look up a state at or before each involved timestamp, and use the unicycle model to predict the value at those timestamps. However, if the relative pose constraints regularly include states older than the lag, then you enter condition (2) above.

Regardless, I think this PR should be merged as-is, and we will pick up this conversation over on the unit test PR #158.

Reply:
:+1: